### ISSUE TITLE
Please create a USER FRIENDLY user interface to set powershell execution policies

### ISSUE THREAD

**Issue by FlightControl-User** [PowerShell/PowerShell#18696]
### Summary of the new feature / enhancement

At the time of writing i own a new microsoft pc for about 1 day.
Powershell restricts all execution of powershell scripts by default.
I need to enable it through a command that is far away hidden.

Please have a setting to allow or configure execution policies of powershell through a user interface or revise the security restrictions. What is the use of security if it blocks completely the funcitonality of the system?

**Comment by StevenBucher98** [comment]
Thanks for submitting this issue @FlightControl-User! In Settings>Privacy & Security > For Developer> PowerShell there is a setting to allow local scripts to run without signing changing the execution policy. This currently only works for Windows PowerShell version 5.1, the version of PowerShell built shipped inbox with Windows. Are you interested in having a similar setting for PowerShell 7?

**Comment by FlightControl-User** [comment]
Thanks for your constructive reply. I would recommend a yes and a no. It's double.
Allow me pls to speak freely. As a simple mind and observer from a distance ...
Powershell is a scripting language and enables to do more things with your computer than it provides out of the box. What good is it to buy a new laptop (from Microsoft), then wanting to use the new laptop using powershell scripts (which worked on my previous laptop and synchronized through one drive on my new local machine), to get blocked by default ???

Also, does bash (linux subsystem) prevent out of the box to run shell scripts ??? I mean, i just feel historically this all has grown as a kind of a mess and needs a revision.

I would suggest to enable remotesigned scripts by default on the machines, and for those who want extra security to prevent or configure extra settings.

And yes, such a UI feature for powershell 7 is needed, or again a lot of complaints will come!
I would recommend Microsoft to think over carefully what the end to end solution and security model is still a fit with modern times ...

**Comment by mklement0** [comment]
@StevenBucher98, while that would be nice, the challenge with respect to PowerShell Core is twofold:
* multiple versions may be installed side by side, which would complicate the UI
* versions may be installed in custom locations that may not be discoverable via `$env:PATH`

@FlightControl-User, I agree that `RemoteSigned` would make for a more sensible default - and in fact it _is_ the default, but only on (recent) _server_ editions of Windows.
(On a more fundamental note, you could argue that simply having _no_ execution policies may be preferable - seems to be no problem for PowerShell Core _on Unix_).

As for:

> I need to enable it through a command that is far away hidden.

It's not _hard_ to discover the relevant command, but it's _harder than it needs to be_:

Currently, the error message you get when a script is blocked is:

```none
'...' cannot be loaded because running scripts is disabled on this system. For more information, 
see about_Execution_Policies at https:/go.microsoft.com/fwlink/?LinkID=135170
```

The link leads to [about_Execution_Policies](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Execution_Policies), and if you're running in _Windows Terminal_ you can even Ctrl-click it directly and it will open in your browser.
That said, given the conceptual nature of the topic, the relevant commands do not appear until much further down the page.

I think it would help if the current error message were _amended_ to include something like the following:

```
For instance, run `Set-ExecutionPolicy -Scope CurrentUser RemoteSigned -Force`
to enable script execution for the current user for all scripts except those downloaded from the web.
```

**Comment by FlightControl-User** [comment]
Hello there,

Thanks for replying. I read your feedback with great interest and thought this over a bit ...
Please consider the below considerations:

> * multiple versions may be installed side by side, which would complicate the UI

@StevenBucher98 , @mklement0 Afraid this is a developer answer. Please think from an end-user perspective. Also, when you buy a car, is a Toyota RAV4 the same as a Toyota Prius? No, they have different dashboards. So yes, the UI may be different depending on the powershell you have on your machine! I'm sorry but complexity comes with design and the battle of backward compatibility needs to be dealt with, considering that this backward compatibility is required.

> It's not _hard_ to discover the relevant command, but it's _harder than it needs to be_:

I would like to emphasize that it is _hard_ to discover the relevant command! An end user needs to jump into a terminal, (imagine, on a user interface driven operating system!). Then, he needs need to magically know the command to set user policies, which indeed appear in the error, but, you need to "see" the error. Sometimes this error appears hidden in an other component and the end user may not even know what is going on. Does Microsoft really want an end user to type in a command like `Set-ExecutionPolicy -Scope CurrentUser RemoteSigned -Force` ???
I spent about half an hour searchin through the quite technical explanations of Microsoft on the Set-ExecutionPolicy which is simply incomprehensible for end users!!! So please think from an end user perspective how the Powershell scripting subsystem can be improved/serviced. What functions would you want to embed in Windows to make it more easy for end users to use the system? Is it only "execution policies" or are there more functions to consider?

> (On a more fundamental note, you could argue that simply having _no_ execution policies may be preferable - seems to be no problem for PowerShell Core _on Unix_).

@mklement0 As stated by myself earlier, I confirm this also, linux, bash has no issue with execution policies at all. So imagine how confusing this all is for end users ... 

> I think it would help if the current error message were _amended_ to include something like the following:

@mklement0 Thank you! That is a good solution, but not **the** solution as an end user. But it will definitely help!

So to summarize:

- Consider from an end user perspective how Powershell configuration can be enhanced within the UI. 
Is execution policy the only thing?
- Consider the -RemoteSigned as the default. Consider even patching this in existing windows systems.
- Consider, as @mklement0 correctly has advised, to correct the error message.

Don't expect me to respond any further to this feedback. I think the message has been written clear enough.

**Comment by kilasuit** [comment]
@FlightControl-User until (if ever) PowerShell 7 is included in Windows, then I don't think you will get a UI baked into Windows for it,

However IMO the **right** place for this kinda feedback would be via the Windows Insider program and it's Feedback hub as this is a Windows UI request (IMO) and not so much PowerShell specific (although the lines are blurry here), so I'd recommend that you raise it there unless others have already done so as I'm not going to.

Also **most** end users will never use PowerShell 7 (or even PowerShell 5.1) so this is perhaps this ask isn't the right one.

re this 
> Does Microsoft really want an end user to type in a command like Set-ExecutionPolicy -Scope CurrentUser RemoteSigned -Force ???

Yes, if they want to learn how to end up being a developer and not just an end user. Because by showing them that and the info in the error message it gives them the choice to go further or not. Which is the right way forward to enabling everyone to reach their true potential. As they say "Curiosity is only a good thing" we should give as many chances as possible to enable people to become more curious.

**Comment by FlightControl-User** [comment]
I've given up on the insider program. It is a time wasting resource hog that brings no direct benefits.
The project manager of the PowerShell team has responded in this request so I assume that he will pick this idea up. I don't think it requires an insider program issue beside the other thousands of issues, and then this request slipping through because nobody reads it.

**Comment by kilasuit** [comment]
@FlightControl-User IMO you are very wrong in your assumption about the insider program and perhaps are expecting changes to occur far too quickly than is ever **realistically** going to happen. Also I was suggesting to you that route as one that you could undertake based on my many years of being very heavily invested in the mass majority of the Microsoft ecosystem, including Windows as an insider since 2014 and until recently the MVP program since 2016.

Whilst its fair to feel like its a waste of time, I think given enough time they will correct some of the things that aren't currently ideal, just like they have always done.
re this 
> I don't think it requires an insider program issue beside the other thousands of issues, and then this request slipping through because nobody reads it.

I think it also needs one as different teams read different feedback mechanisms so being able to link them to get gives much more weight, in my experience 🙂
